media-sound/pulseaudio: Bump to 14.99.1 snapshot and switch to Meson.

pinkflames opened the pull request:
This PR is based on the [pulseaudio-9999.ebuild](https://bugs.gentoo.org/749357#c38) from bugs.gentoo.org as compiled by @joakim-tjernlund from initial work of Igor V. Kovalenko (et al?) who in-turn used the original Gentoo ebuild.

General TODO before this could be considered for merging:
 - [ ] Wire up gstreamer IUSE (WIP)
 - [x] libc RDEPEND needs [polishing up and a check](https://github.com/gentoo/gentoo/pull/20388#discussion_r613810066) on whether PCRE2 is still unsupported
 - [x] Examine and polish multilib
 - [x] Consider offering tdb as a gdbm alternative
 - [ ] Deal with the licensing mess (bug open on b.g.o)
 - [x] ~~Verify that elogind IUSE is not feasible and make D-Bus non-optional, too~~
 - [x] Look into splitting building of libs from the daemon ~~or at least disable autospawn on systemd~~
 - [x] Either document ~~or change pa_meson_multilib_native_use_feature()~~
 - [x] Double check that --localstatedir="${EPREFIX}"/var is really needed in multilib_src_configure
 - [ ] Decide on what the recomended pre-allocated sound buffer size (if any) should be
 - [ ] When done, audit inherit for any unused eclass includes
 - [ ] @leio wants a news item about forget-missing (name provisional) USE flag so that past and current GNOME users can opt-in for database pruning of old routes
 - [ ] prepend EPREFIX before get_udevdir function (autotools ebuild does that)

thesamesam reviewed (commented):
(Slightly outdated comments but I’ll post anyway)

pinkflames commented on media-sound/pulseaudio/pulseaudio-14.2.ebuild:
```diff
@@ -0,0 +1,318 @@
+	local emesonargs=(
+		-Dadrian-aec=false # Not packaged?
+		--localstatedir="${EPREFIX}"/var
+		-Dmodlibexecdir="${EPREFIX}"/usr/"$(get_libdir)"/${P}
```
Good catch, I'll fix that right now. A bit surprising that repoman only caught the EPREFIX.

thesamesam replied:
We don't need to quote ${P} but it looks better than quoting twice for ${EPREFIX} and $(get_libdir).

pinkflames commented on media-sound/pulseaudio/pulseaudio-14.2.ebuild:
```diff
@@ -0,0 +1,318 @@
+		$(meson_use gnome stream-restore-clear-old-devices) # TODO: Get ACK'ed on this
+		$(meson_feature glib)
+		$(meson_feature asyncns)
+		#$(meson_use cpu_flags_arm_neon neon-opt)
```
I'm aware of the thread in one of the Gentoo bugs - I'll tackle that once this ebuild is closer to not being a draft. For what it's worth, that's the upstream recommended solution - migrate once, cry once.

pinkflames replied:
Based on input from @leio renamed to forget-missing (now that I think about it, might need another rename), so the underlying issue of re-using a common USE flag has been averted. We'll also need a news item, which I'll add to the TODO.

pinkflames commented on media-sound/pulseaudio/pulseaudio-14.2.ebuild:
```diff
@@ -0,0 +1,318 @@
+		fi
+	else
+		if ! use elibc_glibc; then
+			# Non-glibc multilib is probably non-existant but just in case:
```
Thank you very much - I knew I made that typo somewhere but somehow had convinced myself I had fixed it already.

thesamesam replied:
It happens ;)

pinkflames commented on media-sound/pulseaudio/pulseaudio-14.2.ebuild:
```diff
@@ -0,0 +1,318 @@
+		fi
+	else
+		# remove foreign abi modules
+		rm -rf "${ED}"/usr/$(get_libdir)/pulse-*/
```
Fixed. Thanks.

scardracs replied:
As told to me lots of time, we don't use the f in rm

pinkflames replied:
Thank you, @scardracs, that's indeed a solid rule to stick with. Fixed in local version. I'll test other things before pushing it.

pinkflames commented on media-sound/pulseaudio/pulseaudio-14.2.ebuild:
```diff
@@ -0,0 +1,318 @@
+	meson_src_install
+
+	if multilib_is_native_abi; then
+		if use doc ; then
```
I was sure I had dealt with all of those spaces before semicolon already. Thanks.

pinkflames commented on media-sound/pulseaudio/pulseaudio-14.2.ebuild:
```diff
@@ -0,0 +1,318 @@
+			use "$1" && echo "-D$define" || echo "-U$define"
+		}
+
+		unifdef $(use_define zeroconf AVAHI) \
```
I'll assume you mean that unifdef as a whole count as an external command rather than the echo which is a shell built-in (since ebuilds are explicitly written in bash, echo would never be an external command, IMO), and I'll commit a fix for that. If I made a mistake, please do explain in greater detail.

pinkflames commented on media-sound/pulseaudio/pulseaudio-14.2.ebuild:
```diff
@@ -0,0 +1,318 @@
+		elog "which to use is done via the 'headset' argument of module-bluetooth-discover."
+	fi
+
+	use equalizer && optfeature "using qpaeq script" dev-python/PyQt5[dbus,widgets]
```
Done.

pinkflames commented on media-sound/pulseaudio/pulseaudio-14.2.ebuild:
```diff
@@ -0,0 +1,318 @@
+
+	if linux_config_exists; then
+		local snd_hda_prealloc_size=$(linux_chkconfig_string SND_HDA_PREALLOC_SIZE)
+		if [ -n "${snd_hda_prealloc_size}" ] && [ "${snd_hda_prealloc_size}" -lt 2048 ]; then
```
Thank you and sorry for not spotting this myself - I probably had too much of a reverence to the ebuild to even consider it might have problematic code in it. The next commit should fix this, let me know if a simple conversion to double brackets is insufficient here.

thesamesam replied:
No worries :)

pinkflames commented on media-sound/pulseaudio/pulseaudio-14.2.ebuild:
```diff
@@ -0,0 +1,318 @@
+}
+
+pa_meson_multilib_native_use_feature() {
+	echo "-D ${2:-${1}}=$(multilib_native_usex ${1} enabled disabled)"
```
If you're okay with it being changed, @Whissi is in the process of picking how pipewire ebuild should handle multilib (also Meson based). If the existing function should be kept, then I'll add an example once I know how to proceed. For now I'll put it on the TODO list for this PR.

thesamesam replied:
Sure.

pinkflames replied:
PipeWire ended up going with this:
`meson_native_feature() {
multilib_native_usex "${1}" "-D${2-${1}}=enabled" "-D${2-${1}}=disabled"
}`
 Personally I don't find it in any way more readable while taking up more bytes. For now I have added a example while renaming the existing helper function implementations to the same as PipeWire ebuild.

pinkflames commented on media-sound/pulseaudio/pulseaudio-14.2.ebuild:
```diff
@@ -0,0 +1,318 @@
+multilib_src_configure() {
+	local emesonargs=(
+		-Dadrian-aec=false # Not packaged?
+		--localstatedir="${EPREFIX}"/var
```
Yes, it does stand out but, if my memory serves me right, when disabled, it put the runtime files under /var/lib/tmp or some other similarly obscure path that's unlikely to get automatically cleaned out. I will make a TODO item to recheck this at some point.

mattst88 replied:
FWIW, we have to do this in `gnome-base/gdm` as well. Their build system seems to assume `localstatedir` should be `/var` rather than `/var/lib`, so this isn't too surprising to me.

thesamesam replied:
Ah, this is where my lack of GNOME-fu matters :)

pinkflames replied:
On retest this is what I see:
    localstatedir:                 /var/lib
...
    System Runtime Path:           /var/lib/run/pulse
    System State Path:             /var/lib/lib/pulse
    System Config Path:            /var/lib/lib/pulse

Most obviously the /var/lib/lib looks immediately weird. Setting --localstatedir=/var removes that lib for more normal looking paths but no idea if they are correct according to any standard that might apply. ;)

pinkflames commented on media-sound/pulseaudio/pulseaudio-14.2.ebuild:
```diff
@@ -0,0 +1,318 @@
+
+EAPI=7
+
+inherit meson bash-completion-r1 flag-o-matic gnome2-utils linux-info optfeature systemd toolchain-funcs udev multilib-minimal
```
TODO item created.

pinkflames commented on media-sound/pulseaudio/pulseaudio-14.2.ebuild:
```diff
@@ -0,0 +1,318 @@
+		if [ -n "${snd_hda_prealloc_size}" ] && [ "${snd_hda_prealloc_size}" -lt 2048 ]; then
+			ewarn "A pre-allocated buffer-size of 2048 (kB) or higher is recommended for the"
+			ewarn "HD-audio driver when using timer-based scheduling!"
+			ewarn "CONFIG_SND_HDA_PREALLOC_SIZE=${snd_hda_prealloc_size}"
```
It's a bit complicated - the new default, I think, is to set it to 0 and have no pre-allocation at all. But, as far as I understand, that also means that there's no guarantee the runtime created buffer will be continuous (or something to that extent). It's further complicated by the fact that presumably we'd like to move towards using PipeWire in the future, which does not currently have the ability to fully utilize buffers even close to that size, leading to wasted memory and allegedly CPU cycles (I guess there's some inherent inefficiency when the application buffers are much smaller than the hardware buffer). For now I have opted to be cautious and not change our kernel configuration recommendation but I more than welcome anyone wishing to express their opinion on this matter to do so either in this PR or perhaps more appropriately on bugs.gentoo.org.

thesamesam replied:
That sounds fine, although it might be worth mentioning this in a comment and tagging the relevant bug(s) to show we've considered it. Thanks for the thought out answer!

pinkflames replied:
I opted to just remove the kernel configuration checks entirely - it's been so long that anyone who cared has long updated their kernel configs and I don't feel like telling people to instead set it to 0 because that would benefit another sound daemon.

pinkflames commented on media-sound/pulseaudio/pulseaudio-14.2.ebuild:
```diff
@@ -0,0 +1,318 @@
+# libpcre needed in some cases, bug #472228
+RDEPEND="
+	|| (
+		elibc_glibc? ( virtual/libc )
```
For now added to the TODO - I'll tackle that with a fresher mind.

thesamesam replied:
No worries. Just giving food for thought.

pinkflames replied:
I'll leave this open for now but I think it's now resolved with just virtual/libc in RDEPEND and dev-libs/libpcre in DEPEND (by accident ended up as BDEPEND for now but I'll fix that in the next push of changes).

joakim-tjernlund commented:
> This PR is based on the [pulseaudio-9999.ebuild](https://bugs.gentoo.org/749357#c38) from bugs.gentoo.org as compiled by @joakim-tjernlund from initial work of Igor V. Kovalenko (et al?) who in-turn used the original Gentoo ebuild.
> 
> General TODO before this could be considered for merging:
> 
>     * [ ]  Wire up gstreamer IUSE (WIP)
> 
>     * [ ]  libc RDEPEND needs [polishing up and a check](https://github.com/gentoo/gentoo/pull/20388#discussion_r613810066) on whether PCRE2 is still unsupported
> 
>     * [x]  Examine and polish multilib
> 
>     * [ ]  Consider offering tdb as a gdbm alternative
> 
>     * [ ]  Deal with the licensing mess (bug open on b.g.o)
> 
>     * [ ]  Verify that elogind IUSE is not feasible and make D-Bus non-optional, too
> 

elogind IUSE is not feasible ? What is that about? Here elogind works fine.

pinkflames commented:
@joakim-tjernlund , I know, I also always get spooked when I read that line - it says elogind **IUSE** is not feasible, because it's currently an automatic dependency, which are not allowed. Per Gentoo guidelines there are exactly two ways to deal with it:

1. Patch the build system, so that it can be *reliably* enabled or disabled via a switch
2. Unconditionally list in RDEPEND and not bother with patching anything

Personally I do not think the anti-elogind people are on the right side of the Linux desktop history, so, while I would accept an upstreamable patch, I have no intention of spending my time on developing one myself. Because D-Bus is always pulled in by elogind, once I have verified that current versions really have no way of easily disabling elogind autodetection, I plan to also remove the D-Bus IUSE, too, since it probably makes no sense to build with elogind enabled but D-Bus disabled.

joakim-tjernlund commented:
> @joakim-tjernlund , I know, I also always get spooked when I read that line - it says elogind **IUSE** is not feasible, because it's currently an automatic dependency, which are not allowed. Per Gentoo guidelines there are exactly two ways to deal with it:
> 
>     1. Patch the build system, so that it can be _reliably_ enabled or disabled via a switch
> 
>     2. Unconditionally list in RDEPEND and not bother with patching anything
> 
> 
> Personally I do not think the anti-elogind people are on the right side of the Linux desktop history, so, while I would accept an upstreamable patch, I have no intention of spending my time on developing one myself. Because D-Bus is always pulled in by elogind, once I have verified that current versions really have no way of easily disabling elogind autodetection, I plan to also remove the D-Bus IUSE, too, since it probably makes no sense to build with elogind enabled but D-Bus disabled.

But elogind is now a build time flag in PA so there should not be a problem or do I misunderstand something  here?

pinkflames commented:
Very interesting, the original 9999 from b.g.o also had those switches - the only reason I can think of for removing them is that perhaps the version I was targeting initially didn't have them yet or they were incorrectly wired up, so I disabled them and made TODO to double check if it's really impossible to reliably disable elogind autodetection. Either way, they're back now. I'll try to pay extra attention wherever they pop up.

joakim-tjernlund commented:
Looked it over and I think the ebuild looks ready now. Could a maintainer commit it now?

pinkflames commented:
Oh, come on, I'm barely half-way through the list of dependencies I extracted from more than 2900 lines of meson.build files, much less identified and removed any stale ones, re-ordered them to at least Gentoo DEPEND order (someone else can have the joy of doing it in GNOME order). And as you might have guessed, I haven't even gone over properly meson_options.txt, because doing so will cause a lot of churn in the diff. And the test failure is still present (maybe it only shows up with the daemon enabled and I never caught it before?).

joakim-tjernlund commented:
> Oh, come on, I'm barely half-way through the list of dependencies I extracted from more than 2900 lines of meson.build files, much less identified and removed any stale ones, re-ordered them to at least Gentoo DEPEND order (someone else can have the joy of doing it in GNOME order). And as you might have guessed, I haven't even gone over properly meson_options.txt, because doing so will cause a lot of churn in the diff. And the test failure is still present (maybe it only shows up with the daemon enabled and I never caught it before?).

I am sure you can find more to tweak but I do think this PA is on pair/better with the current PA-13 in Gentoo

perfect7gentleman commented:
the ebuild for 14.99.1 is so bad.
```
meson_native_true() {
	if multilib_is_native_abi; then
		echo "-D${1}=true"
	else
		echo "-D${1}=false"
	fi
}
```
make all TRUE if even they are disabled.
```
meson setup --buildtype plain --libdir lib64 --localstatedir /var/lib --prefix /usr --sysconfdir /etc --wrap-mode nodownload --build.pkg-config-path /usr/share/pkgconfig --pkg-config-path /usr/share/pkgconfig --native-file /tmp/portage/media-sound/pulseaudio-14.99.1/temp/meson.x86_64-pc-linux-gnu.amd64.ini -Dadrian-aec=false --localstatedir=/var -Dmodlibexecdir=/usr/lib64/pulseaudio-14.99.1 -Dudevrulesdir=/lib/udev/rules.d -Dbashcompletiondir=/usr/share/bash-completion/completions -Dalsa=disabled -Dbluetooth=true -Ddaemon=true -Dnative-headset=true -Dofono-headset=true -Dgsettings=enabled -Dgstreamer=disabled -Dgtk=disabled -Djack=disabled -Dsamplerate=disabled -Dlirc=disabled -Dorc=disabled -Doss-output=disabled -Dopenssl=disabled -Dtest=true -Dudev=disabled -Dwebrtc-aec=disabled -Davahi=disabled -Dfftw=disabled -Dsoxr=disabled -Ddatabase=simple -Dstream-restore-clear-old-devices=false -Dglib=enabled -Dasyncns=disabled -Dtcpwrap=disabled -Ddbus=disabled -Delogind=disabled -Dx11=disabled -Dsystemd=disabled -Dipv6=false -Dpulsedsp-location=/usr/\\$$LIB/pulseaudio /tmp/portage/media-sound/pulseaudio-14.99.1/work/pulseaudio-14.99.1 /tmp/portage/media-sound/pulseaudio-14.99.1/work/pulseaudio-14.99.1-abi_x86_64.amd64
```
```
[ebuild     U  ] media-sound/pulseaudio-14.99.1::pg_overlay [14.2_p20210423::pg_overlay] USE="glib -X -alsa -alsa-plugin -asyncns -bluetooth -daemon -dbus -doc -elogind -equalizer -gdbm -gnome -gstreamer -gtk -ipv6 -jack -lirc -native-headset -ofono-headset -orc (-oss) (-selinux) -sox -ssl (-system-wide) -systemd -tcpd -test -udev -webrtc-aec -zeroconf" 0 KiB
```

mattst88 commented:
> the ebuild for 14.99.1 is so bad.

Try giving feedback without being a dick.

i-garrison commented on media-sound/pulseaudio/pulseaudio-14.99.1.ebuild:
```diff
@@ -0,0 +1,386 @@
+		-Dudevrulesdir="$(get_udevdir)"/rules.d
+		-Dbashcompletiondir="$(get_bashcompdir)" # Alternatively DEPEND on app-shells/bash-completion for pkg-config to provide the value
+		$(meson_native_feature alsa)
+		$(meson_native_true bluetooth bluez5)
```
this should account for actual use flags via `multilib_native_usex` just like meson_native_feature does, and probably be named like `meson_native_use_enable`
as far as I can see currently it is just `true` for native ABI which is not what we want

pinkflames replied:
I know this's a long thread but there already was a [discussion on that](https://github.com/gentoo/gentoo/pull/20388#discussion_r613812151) right here as well as talks about the same topic regarding another package with a different Gentoo developer privately, and they unanimously did not approve of any kind of ebuild specific prefix to custom functions.

i-garrison replied:
`foo_boolean` is more than fine :) main point is that first branch in this helper should be something like this
```
foo_boolean() {
    local use_flag="${1}"
    local meson_option="${2:-${1}}"
    
    	if multilib_is_native_abi; then
    		echo "-D${meson_option}=$(multilib_native_usex ${use_flag} true false)"
    	else
    		echo "-D${meson_option}=false"
    	fi
}
```
i.e. not just return `true` but actually follow use flag

I'm not sure if it actually makes sense to disable meson options for non-native ABI this way, maybe only for disabling `daemon` for non-native ABI?
Initially I opted to employ `multilib_native_usex` to let non-native ABI follow use flags as well, so maybe there should be two helpers, `foo_native_abi_boolean` and `foo_boolean` ?

pinkflames replied:
Oh, I see what you mean, I think something may have gotten mixed up, when I ported those functions from the PipeWire jack-sdk & multilib PR. I'll think some more about this, before making any commits addressing it, but it looks at least the daemon one might, indeed, be behaving incorrectly. Thanks for noticing.

i-garrison replied:
Correct, just looked at pipewire ebuild - it is using `meson_native_enabled` function to unconditionally enable meson features for native abi, note those meson features of pipewire do not depend on use flags.

i-garrison replied:
Hi @pinkflames I see packages are now switching to `meson-multilib` eclass which provides `meson_native_use_bool` `meson_native_use_feature` and `meson_native_enabled` already, no need to define these in the ebuild. `pipewire` ebuild is already converted.

pinkflames replied:
Thanks for the heads up. I actually converted my personal pipewire fork to meson-multiib.eclass just days after it became available. The pulseaudio ebuild hasn't been converted yet but it's the next thing I'll do.

joakim-tjernlund commented:
Seems a bit stuck now?

pinkflames commented:
@joakim-tjernlund, I have bumped it to 14.99.1_p20210607 locally but it's not yet at a point where there's any meaningful ebuild improvements to push out but once I do, it will be using meson-multilib.eclass and hopefully we'll be much closer to merging.

i-garrison commented on media-sound/pulseaudio/pulseaudio-14.99.1_p20210607.ebuild:
```diff
@@ -0,0 +1,351 @@
+		$(meson_feature X x11)
+		$(meson_feature systemd)
+		$(meson_use ipv6)
+	)
```
for recent snapshot please add this
`$(meson_native_use_feature doc doxygen)`

pinkflames replied:
Right, I bumped it before diffing the build systems, since I wanted to do that after finishing dependency checkup for the old build. Sorry for missing it, should land in the next push, if not, be sure to remind me (but hopefully I won't forget, since it should land with that).

i-garrison commented on media-sound/pulseaudio/pulseaudio-14.99.1_p20210607.ebuild:
```diff
@@ -0,0 +1,351 @@
+		$(meson_feature glib) # WARNING: toggling this likely changes ABI
+		$(meson_feature asyncns)
+		#$(meson_use cpu_flags_arm_neon neon-opt)
+		$(meson_feature tcpd tcpwrap) # TODO: system-wide specific?
```
`tcpwrap` only affects daemon so this should be `meson_native_use_feature` as well

pinkflames replied:
Indeed, I wonder if this is a change since previous version or if I missed it all this time. Either way will get fixed with the next push. Thank you for noticing and reporting this.

Whissi commented:
Thank you all for your work! I pushed the initial version with some adjustments as pre release without keywords to allow further testing. Let's continue the work based on pulseaudio-14.2_pre ebuild in new bug/PR.
